pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/ConsumptionDelayTracker.java

```diff
@@ -0,0 +1,361 @@
+ *    _________|__________                \
+ *   | TimerTrackingTask |          (CONSUMING -> DROPPED state change)
+ *   |___________________|
+ *
+ */
+
+public class ConsumptionDelayTracker {
+
+  // Sleep interval for timer thread that triggers read of ideal state
+  private static final int TIMER_THREAD_TICK_INTERVAL_MS = 300000; // 5 minutes +/- precision in timeouts
+  // Once a partition is marked for verification, we wait 10 minutes to pull its ideal state.
+  private static final int PARTITION_TIMEOUT_MS = 600000;          // 10 minutes timeouts
+  // Delay Timer thread for this time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 100;
+
+  /*
+   * Class to keep a Pinot Consumption Delay measure and the time when the sample was taken (i.e. sample time)
+   * We will use the sample time to increase consumption delay when a partition stops consuming: the time
+   * difference between the sample time and current time will be added to the metric when read.
+   */
+  static private class DelayMeasure {
+    public DelayMeasure(long t, long d) {
+      _delayMilliseconds = d;
+      _sampleTime = t;
+    }
+    public long _delayMilliseconds;
+    public long _sampleTime;
+  }
+
+  /*
+   * Class to handle timer thread that will track inactive partitions
+   */
+  private class TrackingTimerTask extends TimerTask {
+    final private ConsumptionDelayTracker _tracker;
+
+    public TrackingTimerTask(ConsumptionDelayTracker tracker) {
+      _tracker = tracker;
+    }
+
+    @Override
+    public void run() {
+      // tick inactive partitions every interval of time to keep tracked partitions up to date
+      _tracker.timeoutInactivePartitions();
+    }
+  }
+
+  // HashMap used to store delay measures for all partitions active for the current table.
+  // _partitionsToDelaySampleMap<Key=PartitionGroupId,Value=DelaySample>
+  private ConcurrentHashMap<Integer, DelayMeasure> _partitionToDelaySampleMap = new ConcurrentHashMap<>();
+  // We mark partitions that go from CONSUMING to ONLINE in _partitionsMarkedForVerification: if they do not
+  // go back to CONSUMING in some period of time, we confirm whether they are still hosted in this server by reading
+  // ideal state. This is done with the goal of minimizing reading ideal state for efficiency reasons.
+  // _partitionsMarkedForVerification<Key=PartitionGroupId,Value=TimePartitionWasMarkedForVerificationMilliseconds>
+  private ConcurrentHashMap<Integer, Long> _partitionsMarkedForVerification = new ConcurrentHashMap<>();
+  // Mutable versions of timer constants so we can test with smaller delays
+  final int _timerThreadTickIntervalMs;
+  // Timer task to check partitions that are inactive against ideal state.
+  final private Timer _timer;
+
+  final private ServerMetrics _serverMetrics;
+  final private String _tableNameWithType;
+
+  private boolean _enableAging;
+  private boolean _enablePerPartitionMetric = true;
```

Comment:
make these final, and don't set the values here

Reply:
Done